# Problem 1

$f(x_1, x_2) = 2x_1^2 - 4x_1x_2+1.5x_2^2+x_2$

Show that the stationary point (zero gradient) of the function is a saddle (with indefinite Hessian).

\begin{gather}
\nabla f(x)  =  
\begin{bmatrix}
4x_1-4x_2 \\
-4x_1+3x_2+1 
\end{bmatrix}
\end{gather}

The gradient is 0 at $x_1 = 1$, $x_2 = 1$.

\begin{gather}
 H  =  
\begin{bmatrix}
4 & -4 \\
-4 & 3\end{bmatrix}
\end{gather}

Eigen values for Hessian are −0.53 and 7.53. Hessian is indefinite and we have a saddle.

Find the directions of downslopes away from the saddle.

Taylor expansion around saddle point can be written as
\begin{gather}
\begin{split}
f(x_1, x_2) &\approx f(1,1) + \left. \nabla_x f\right|_{x_0}^T(x-x_0) + \frac{1}{2} (x-x_0)^T \left. H_f\right|_{x_0}(x-x_0)   
\end{split}
\end{gather}

Since

$f(1,1) = \frac{1}{2}, \left. \nabla_x f\right|_{x_0} = 0 \text{    and    } H = \begin{bmatrix}
4 & -4 \\
-4 & 3\end{bmatrix}$

\begin{gather}
\begin{split}
f(x_1, x_2) &\approx \frac{1}{2} + \frac{1}{2} [x_1-1 \quad x_2-1] \begin{bmatrix}
4 & -4 \\
-4 & 3\end{bmatrix}
\begin{bmatrix}
x_1-1  \\
x_2-1 \end{bmatrix} \\
f(x_1, x_2) - \frac{1}{2} &\approx 2(x_1-1)^2 - 4(x_1-1)(x_2-1)+\frac{3}{2}(x_2-1)^2
\end{split} 
\end{gather}

This can be rewritten in the form of $[a(x_1-1)-b(x_2-1)][c(x_1-1)-d(x_2-1)]$.<br>
And the roots are $a = 3$, $b = \frac{3}{2}$, $c = \frac{2}{3}$ and $d = 1$


Then, 
\begin{gather}
f(x_1, x_2) - \frac{1}{2} &\approx [3(x_1-1)-\frac{3}{2}(x_2-1)][\frac{2}{3}(x_1-1)-(x_2-1)]
\end{gather}

For $f(x_1, x_2) - \frac{1}{2} < 0 $, $[3(x_1-1)-\frac{3}{2}(x_2-1)][\frac{2}{3}(x_1-1)-(x_2-1)] < 0$

\begin{gather}
[3(x_1-1)-\frac{3}{2}(x_2-1)] < 0 \quad\text{or} \quad [\frac{2}{3}(x_1-1)-(x_2-1)]<0
\end{gather}

\begin{gather}
3\partial x_1-\frac{3}{2}\partial x_2 <0 \quad \text{or} \quad \frac{2}{3}\partial x_1-\partial x_2 <0 \\
\end{gather}

The direction of downslopes away from saddle is

\begin{gather}
2\partial x_1 <\partial x_2 \quad \text{or} \quad 2\partial x_1 <3\partial x_2 \\
\end{gather}

 <br /><br />  <hr /> <br /> <br />

# Problem 2 (a)

 Find the point in the plane $x_1 + 2x_2 + 3x_3 = 1$ in $R^3$ that is nearest to the point $(−1, 0, 1)^T$. Is this a convex problem?

The problem can be defined using distance formula as:

\begin{gather}
\text{min}\quad d^2 = (x_1+1)^2+(x_2)^2+(x_3-1)^2
\end{gather}

Then, $f(x_1, x_2, x_3) = x_1^2 + 2x_1+ 1 + x_2^2+x_3^2-2x_3+1 $

$f(x_1, x_2, x_3) = x_1^2 + 2x_1 + x_2^2+x_3^2-2x_3+ 2 $

\begin{gather}
\nabla f = 
\begin{bmatrix}
2x_1+2 \\
2x_2 \\
2x_3-2 \\
\end{bmatrix}
\end{gather}

\begin{gather}
H = 
\begin{bmatrix}
2& 0& 0\\
0& 2& 0\\
0& 0& 2\\
\end{bmatrix}
\end{gather}

Eigen values for this Hessian is $\lambda_1, \lambda_2, \lambda_3 = 2$

From $x_1 + 2x_2 + 3x_3 = 1$, $x_1 = 1 - 2x_2 - 3x_3$

Then, $f(x_2, x_3) = (1 - 2x_2 - 3x_3)^2+ 2x_1 + x_2^2+x_3^2-2x_3+ 2$

$f(x_2, x_3) = 5x_2^2 + 12x_2x_3 - 8x_2 + 10x_3^2 - 14x_3+ 15$

\begin{gather}
\nabla f = 
\begin{bmatrix}
10x_2+12x_3-8 \\
12x_2+20x_3-14 \\
\end{bmatrix}
\end{gather}

\begin{gather}
H = 
\begin{bmatrix}
10& 12\\
12& 20\\
\end{bmatrix}
\end{gather}

Eigen values for this Hessian is $\lambda_1 = 28, \lambda_2 = 2$

To find the closest point, $\nabla f$ is set to 0

\begin{gather}
10x_2+12x_3-8 = 0 \\
12x_2+20x_3-14 = 0
\end{gather}

The solution is:

$x_2 = \frac{-1}{7}, x_3 = \frac{11}{14}, x_1 = \frac{-15}{14}$

This is a convex problem because the hessian is positive definite.

# Problem 2 (b)

## Part 1: Gradient Descent

In [ ]:
import numpy as np
import matplotlib.pyplot as plt

Problem definition:

In [ ]:
def func(x):
    x2 = x[0]
    x3 = x[1]
    return 5 * x2 ** 2 + 12 * x2 * x3 - 8 * x2 + 10 * x3 ** 2 - 14 * x3 - 5

Gradient of the problem:

In [ ]:
def grad_func(x):
    x2 = x[0]
    x3 = x[1]
    g = np.array([10 * x2 + 12 * x3 - 8, 12 * x2 + 20 * x3 - 14])
    return g

In [ ]:
def phi_func(alpha, t, f, g):
    phi = f - t*np.matmul(g.T,g)*alpha 
    return phi

In [ ]:
def armijo(x, alpha = 1, t = 0.5, max_iter = 1000):
    iteration = 0
    f = func(x)
    g = grad_func(x)
    f_shift = func(x-alpha*g)
    Phi = phi_func(alpha, t, f, g)
    
    while f_shift>Phi and iteration<max_iter:
        f = func(x)
        g = grad_func(x)
        f_shift = func(x-alpha*g)
        Phi = phi_func(alpha, t, f, g)
        alpha = 0.5*alpha
        iteration += 1

    return alpha 

In [ ]:
def gradient_descent(x, alpha = 1, t = 0.5, max_iter = 1000):
    summary = []
    iteration = 0;
    f = func(x)
    g = grad_func(x)
    
    while iteration<max_iter:
        alpha = armijo(x,alpha,t,max_iter)
        x = x - alpha*g
        f = func(x)
        g = grad_func(x)
        iteration += 1
        summary.append([x[0],x[1], f,alpha, iteration])
    return x, summary

In [ ]:
x, summary = gradient_descent(np.array([[0],[0]]),1,0.5,1000)

In [ ]:
x2 = x[0]
print(x2)
x3 = x[1]
print(x3)
x1 = 1- 2*x2 - 3*x3
print(x1)

In [ ]:
def extract(lst, index):
    return [item[index] for item in lst]

In [ ]:
f = extract(summary, 2)
iteration = extract(summary, 4)

Testing another initial point

In [ ]:
x, summary = gradient_descent(np.array([[1],[-1]]),1,0.5,1000)
x2 = x[0]
print(x2)
x3 = x[1]
print(x3)
x1 = 1- 2*x2 - 3*x3
print(x1)
f1 = extract(summary, 2)
iteration1 = extract(summary, 4)

(1) Initial points tested are [0,0] and [1,-1]. Regardless of the initial points, algorithm reaches to the points which minimize the distance.
(2) Solution for initial point [0,0] is [-1.07142857, -0.14285695, 0.78571416] and solution for initial point [1,-1] is [-1.07142857 -0.14285703, 0.78571421].

In [ ]:
plt.xscale("log")
plt.xlabel("iterations")
plt.ylabel("cost function")
plt.plot(iteration, f, label = "initial point [0,0]")
plt.plot(iteration1, f1, label = "initial point [1,-1]")
plt.legend(loc="upper right")

## Part 2: Newton's method

In [ ]:
def Newton(x, max_iter = 100, step = 0.5):
    summary = []
    H = np.array([[10, 12],[12, 20]])
    iteration = 0
    f = func(x)
    g = grad_func(x)
    g_norm = np.linalg.norm(g)
    while iteration < max_iter: 
        x = x - step*np.matmul(np.linalg.inv(H),g)
        f = func(x)
        g = grad_func(x)
        g_norm = np.linalg.norm(g)
        iteration += 1;
        summary.append([x[0],x[1], f, iteration])
    return x, summary

In [ ]:
x, summary = Newton(np.array([[0],[0]]),100)

In [ ]:
x2 = x[0]
print(x2)
x3 = x[1]
print(x3)
x1 = 1- 2*x2 - 3*x3
print(x1)

In [280]:
f = extract(summary, 2)
iteration = extract(summary, 3)
x, summary = Newton(np.array([[1],[-1]]),100)
x2 = x[0]
print(x2)
x3 = x[1]
print(x3)
x1 = 1- 2*x2 - 3*x3
print(x1)
f1 = extract(summary, 2)
iteration1 = extract(summary, 3)

[-0.14285714]
[0.78571429]
[-1.07142857]


(1) Initial points tested are [0,0] and [1,-1]. Regardless of the initial points, algorithm reaches to the points which minimize the distance.
(2) Solution for initial point [0,0] is [-1.07142857, 0.78571429, -0.14285714] and solution for initial point [1,-1] is [-1.07142857, 0.78571429, -0.14285714].

In [ ]:
plt.xscale("log")
plt.xlabel("iterations")
plt.ylabel("cost function")
plt.plot(iteration, f, label = "initial point [0,0]")
plt.plot(iteration1, f1, label = "initial point [1,-1]")
plt.legend(loc="upper right")

 <br /><br />  <hr /> <br /> <br />

# Problem 3

Prove that a hyperplane is a convex set. Hint: A hyperplane in $R^n$ can be expressed as: $a^Tx = c$ for $x ∈ R^n$, where a is the normal direction of the hyperplane and c is some constant.


For arbitrary points $x_1$ and $x_2$ where $x_1, x_2 \in S$ 

\begin{gather}
a^Tx_1 = c \\
a^Tx_2 = c
\end{gather}

The space is convex if $\lambda x_1+(1-\lambda)x_2 \in S$

Linear transformation by $a$ gives:

\begin{gather}
a^T(\lambda x_1+(1-\lambda)x_2) = \lambda a^Tx_1+(1-\lambda)a^Tx_2 = \lambda c+(1-
\lambda)c = c
\end{gather}

The hyperplane is convex.

 <br /><br />  <hr /> <br /> <br />

# Problem 5

Let c(x) be the cost of producing x amount of product A and assume that c(x) is differentiable everywhere. Let y be the price set for the product. Assuming that the product is sold out. The total profit is defined as: $c^*(y) = \underset{x}{\text{max}}\{xy-c(x)\}$


Show that $c^∗(y)$ is a convex function with respect to y.

We need to show that $xy-c(x)$ is convex for c^* to be convex.

\begin{gather}
\nabla_y f = x \\
H = 0
\end{gather}

Since the hessian is positive semi definite, the space is the convex set.